# Working with titanic dataset

import all library that are going to be used

In [1]:
import pandas as pd
import numpy as np

import titanic dataset with tab saparator, and try to display sample data

In [2]:
data = pd.read_csv("titanic.tsv", sep='\t')
data.head()

PassengerId  Survived  Pclass  \
0            1         0     3.0   
1            2         1     1.0   
2            3         1     3.0   
3            4         1     1.0   
4            5         0     3.0   

                                                Name     Sex Age  SibSp Parch  \
0                            Braund, Mr. Owen Harris    male  22      1     0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38      1     0   
2                             Heikkinen, Miss. Laina  female  26      0     0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35      1     0   
4                           Allen, Mr. William Henry    male  35      0     0   

             Ticket     Fare Cabin Embarked     ship  
0         A/5 21171     7,25   NaN        S  Titanic  
1          PC 17599  71,2833   C85        C  Titanic  
2  STON/O2. 3101282    7,925   NaN        S  Titanic  
3            113803     53,1  C123        S  Titanic  
4            373450     8,05   NaN        S  Titanic

check datatypes to check if the data is in correct types

In [3]:
data.dtypes

PassengerId      int64
Survived         int64
Pclass         float64
Name            object
Sex             object
Age             object
SibSp            int64
Parch           object
Ticket          object
Fare            object
Cabin           object
Embarked        object
ship            object
dtype: object

from dtypes function we know that Pclass, Age, Parch and Fare column are in object type, so we should change it to numeric datatype

In [4]:
data['Pclass'] = pd.to_numeric(data.Pclass, errors='coerce')
data['Age'] = data['Age'].str.replace(',','.')
data['Age'] = pd.to_numeric(data.Age, errors='coerce')

data['Parch'] = pd.to_numeric(data.Parch, errors='coerce')

data['Fare'] = data['Fare'].str.replace(',','.')
data['Fare'] = pd.to_numeric(data.Fare, errors='coerce')

data.dtypes

PassengerId      int64
Survived         int64
Pclass         float64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch          float64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
ship            object
dtype: object

We will try to describe data to understand it

In [5]:
data.describe()

PassengerId    Survived      Pclass          Age       SibSp  \
count   894.000000  894.000000  894.000000   721.000000  894.000000   
mean    446.030201    0.381432    2.305369    35.836019    0.604027   
std     259.208003    0.508529    0.847653   164.927968    2.571231   
min     -12.000000   -4.000000   -2.000000   -12.000000    0.000000   
25%     223.250000    0.000000    2.000000    20.000000    0.000000   
50%     444.500000    0.000000    3.000000    28.000000    0.000000   
75%     668.750000    1.000000    3.000000    38.000000    1.000000   
max    1143.000000    1.000000    3.000000  4435.000000   70.000000   

            Parch        Fare  
count  893.000000  892.000000  
mean     0.371781   32.075985  
std      0.768325   49.868844  
min      0.000000  -90.000000  
25%      0.000000    7.895800  
50%      0.000000   14.454200  
75%      0.000000   30.771850  
max      5.000000  512.329200

Locate Missing Data using isnull function

In [6]:
data.isnull()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   
891        False     False   False  False  False  False  False  False   False   
892        False     False   False  False  False  False  False  False   False   
893        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked   ship  
0    False   True     False  False  
1    False  False     False  False  
2    False   True     False  False  
3    False  False     False  False  
4    False   True     False  False  
..     ...    ...       ...    ...  
889  False  False     False  False  
890  False   True     False  False  
891  False  False     False  False  
892  False  False     False  False  
893  False   True     False  False  

[894 rows x 13 columns]

expand the missing data of each feature, from below result we know there are 6 attributes with missing value (Age, Parch, Ticket, Fare, Cabin, Embarked)

In [7]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            173
SibSp            0
Parch            1
Ticket           1
Fare             2
Cabin          686
Embarked         2
ship             0
dtype: int64

## Missing values

There are two basic options regarding missing data. 1) Drop the data or, 2) Input missing data.

### Cabin

Because we have 894 rows and there are 686 missing values in Cabin, so we can drop it.

In [8]:
data.drop("Cabin", inplace =True, axis =1)

In [9]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0     3.0   
1            2         1     1.0   
2            3         1     3.0   
3            4         1     1.0   
4            5         0     3.0   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked     ship  
0    0.0         A/5 21171   7.2500        S  Titanic  
1    0.0          PC 17599  71.2833        C  Titanic  
2    0.0  STON/O2. 3101282   7.9250        S  Titanic  
3    0.0            113803  53.1000        S  Titanic  
4    0.0            373450   8.0500        S  Titanic

### Check for symmetric data

In [10]:
data.skew(axis=0)

C:\Users\AndikaBram\AppData\Local\Temp\ipykernel_5600\3606978425.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  data.skew(axis=0)


PassengerId     0.023433
Survived       -0.280494
Pclass         -0.737586
Age            26.432551
SibSp          22.338024
Parch           2.523859
Fare            4.724161
dtype: float64

### Age

We can input missing age data with the median value of the age

In [11]:
data['Age'].fillna(np.nanmedian(data['Age']), inplace = True)

### Parch

We can input missing parch data with the mean value of the parch because the parch data is symmetrical

In [12]:
data['Parch'] = data['Parch'].fillna(np.mean(data['Parch']))

### Ticket

We can input missing ticket data with the mode value of the ticket

In [13]:
data['Ticket'].fillna(data['Ticket'].mode()[0], inplace=True)

### Fare

We can input missing fare data with the mean value of the fare because the fare data is symmetrical

In [14]:
data['Fare'] = data['Fare'].fillna(np.mean(data['Fare']))

### Embarked

We can input missing embarked data with the mode value of the embarked

In [15]:
data['Embarked'].fillna(data['Embarked'].mode()[0], inplace=True)

Check missing value

In [16]:
data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
ship           0
dtype: int64

## Checking outliers

### Survived

In [17]:
data['Survived'].value_counts()

 0    548
 1    345
-4      1
Name: Survived, dtype: int64

We can see that there are survived with value -4. Hence, because it's survived so the value should be 0 or 1, so we change it with mode of survived.

In [18]:
data['Survived'] = data['Survived'].replace([-4], data['Survived'].mode()[0])

### Pclass

In [19]:
data['Pclass'].value_counts()

 3.0    493
 1.0    216
 2.0    184
-2.0      1
Name: Pclass, dtype: int64

We can see that there are one Pclass data with value -2. So it should be the outlier, because there should be 3 Pclass (1, 2, 3) in titatic ship. We can change the value from -2 to 2, because it can be a typo while data was collected.

In [20]:
data['Pclass'] = data['Pclass'].replace([-2], 2)

### Sex

In [21]:
data['Sex'].value_counts()

male       575
female     310
fem          2
malef        1
mal          1
femmale      1
feemale      1
Female       1
malee        1
F            1
Name: Sex, dtype: int64

We can see that thare are some different entry for sex, so we will change it to only male and female. female(female, fem, femmale, feemale, Female, F) male(male, malef, mal, malee)

In [22]:
data['Sex'] = data['Sex'].replace("malef", "male")
data['Sex'] = data['Sex'].replace("mal", "male")
data['Sex'] = data['Sex'].replace("malee", "male")

data['Sex'] = data['Sex'].replace("fem", "female")
data['Sex'] = data['Sex'].replace("femmale", "female")
data['Sex'] = data['Sex'].replace("feemale", "female")
data['Sex'] = data['Sex'].replace("Female", "female")
data['Sex'] = data['Sex'].replace("F", "female")

### Age

In [23]:
data["Age"].describe()

count     894.000000
mean       34.319653
std       148.125488
min       -12.000000
25%        22.000000
50%        28.000000
75%        35.000000
max      4435.000000
Name: Age, dtype: float64

We can see that for age the maximum value is 4435, minimum value is -12, and there is bigger value of std so there are outlier in Age.

In [24]:
data['Age'].value_counts()

28.0    197
24.0     30
22.0     27
18.0     26
19.0     25
       ... 
12.0      1
70.5      1
66.0      1
0.9       1
0.7       1
Name: Age, Length: 94, dtype: int64

We can change the outlier value with median value of Age. We devine the age range is from 1 - 150.

In [25]:
data.loc[data['Age'] > 150, 'Age'] = np.nanmedian(data['Age'])
data.loc[data['Age'] < 1, 'Age'] = np.nanmedian(data['Age'])

In [26]:
print(data.loc[[31]])

    PassengerId  Survived  Pclass  \
31           32         1     1.0   

                                              Name     Sex   Age  SibSp  \
31  Spencer, Mrs. William Augustus (Marie Eugenie)  female  28.0      1   

    Parch    Ticket      Fare Embarked     ship  
31    0.0  PC 17569  146.5208        C  Titanic  


In [27]:
data['Age'].value_counts()

28.0    210
24.0     30
22.0     27
18.0     26
19.0     25
       ... 
12.0      1
36.5      1
55.5      1
23.5      1
74.0      1
Name: Age, Length: 83, dtype: int64

### SibSp

In [28]:
data['SibSp'].value_counts()

0     606
1     213
2      28
4      18
3      16
8       7
5       5
70      1
Name: SibSp, dtype: int64

SibSp is the number of siblings or spouse of a person onboard. So 70 and 8 is outlier, we will change it with mode of SibSp

In [29]:
data['SibSp'] = data['SibSp'].replace([70], data['SibSp'].mode()[0])
data['SibSp'] = data['SibSp'].replace([8], data['SibSp'].mode()[0])

In [30]:
data['SibSp'].value_counts()

0    614
1    213
2     28
4     18
3     16
5      5
Name: SibSp, dtype: int64

### Parch

In [31]:
data['Parch'].value_counts()

0.000000    679
1.000000    121
2.000000     80
3.000000      5
5.000000      4
4.000000      4
0.371781      1
Name: Parch, dtype: int64

Parch is the number of parents / children aboard the Titanic. So 28 is outlier, we will change it with mode of Parch

In [32]:
data['Parch'] = data['Parch'].replace([28.0], data['Parch'].mode()[0])

### Fare

In [33]:
data['Fare'].describe()

count    894.000000
mean      32.075985
std       49.812969
min      -90.000000
25%        7.903100
50%       14.454200
75%       31.000000
max      512.329200
Name: Fare, dtype: float64

We will change fare outlier(value with number below 0) with mean value of fare.

In [34]:
data.loc[data['Fare'] < 1, 'Fare'] = np.mean(data['Fare'])

### Embarked

In [35]:
data['Embarked'].value_counts()

S     647
C     167
Q      76
So      2
Co      1
Qe      1
Name: Embarked, dtype: int64

We can see that thare are some different entry for Embarked, so we will change it

In [36]:
data['Embarked'] = data['Embarked'].replace("So", "S")
data['Embarked'] = data['Embarked'].replace("Co", "C")
data['Embarked'] = data['Embarked'].replace("Qe", "Q")

## Result

In [37]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   894.000000  894.000000  894.000000  894.000000  894.000000   
mean    446.030201    0.385906    2.309843   29.495526    0.463087   
std     259.208003    0.487081    0.835370   12.829530    0.879200   
min     -12.000000    0.000000    1.000000    1.000000    0.000000   
25%     223.250000    0.000000    2.000000   22.000000    0.000000   
50%     444.500000    0.000000    3.000000   28.000000    0.000000   
75%     668.750000    1.000000    3.000000   35.000000    1.000000   
max    1143.000000    1.000000    3.000000   80.000000    5.000000   

            Parch        Fare  
count  894.000000  894.000000  
mean     0.371781   32.750723  
std      0.767895   49.466207  
min      0.000000    4.012500  
25%      0.000000    7.925000  
50%      0.000000   15.245800  
75%      0.000000   32.075985  
max      5.000000  512.329200

### Saving new data

In [38]:
data.to_csv('titanic_clean.csv')